### PR TITLE
Adds Acceptance form options to Asset page: Reminders and Resend

### PR DESCRIPTION
**Pull request by Godmartinz**
# Description
This adds Acceptance Form options. Which include sending another email reminding the user to sign the EULA, and resetting the signature and pending status for a better capture of penmanship.
image
image

The reset EULAs will be available in the same place under "Accept Assets"
image

Fixes # sc-17650 #8818 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* PHP version:
* MySQL version
* Webserver version
* OS version

# Checklist:

- [ ] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [ ] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by dampfklon** [comment]
This duplicates funtionality from the [unaccepted assets report](https://demo.snipeitapp.com/reports/unaccepted_assets) in a as far as i can tell incompatible way (generating a new Acceptance Request instead sending a reminder for the exsiting one)
Would be nice if this coud be done in a unifed way.

Edit: On a second look this does not create a new Acceptance Request but resets an already singed one, I dont quite get the use case for this function and the button does not indicate this potentialy "destuctive" behaivor

**Comment by Godmartinz** [comment]
@dampfklon So the use case for this button isn't to remind users to sign. It is a redo button. its intended to fix bad signatures--a dot perhaps or "daffy duck".  As for the destructiveness, all it would take to fix anything is another signature from the current user. But im considering a modal or warning perhaps. Thanks for your input.

**Comment by dampfklon** [comment]
If this is indeed the usecase i would suggest renaming the button `Resend acceptance form` to something like `Request acceptance resign` which makes it clear. Currently i would assume this button would trigger the reminder action.

I was mislead as the linked issue #8818 also talked about reminders.

File: resources/views/modals/redo-eula.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<!-- Modal -->
+<div class="modal fade" id="redoEula" tabindex="-1" role="dialog" aria-labelledby="redoEulaLabel" aria-hidden="true">
+    <div class="modal-dialog">
+        <div class="modal-content">
+            <div class="modal-header">Send...</div>
+                <div class="modal-body">
+                    <h5>Remind User to Sign</h5>
```
Comment (snipe) [inline resources/views/modals/redo-eula.php]: This would need a form tag here - it performs an action, and all interactions that cause data to change or cause an event to happen need to be in a POST, not a link or a GET, FYI :)

File: app/Http/Controllers/Account/AcceptanceController.php [hunk 2]
```diff
@@ -250,4 +238,50 @@ public function store(Request $request, $id)
...
+            ->latest('created_at')
+            ->get()
+            ->first();
+
+
+        if( ( !$acceptance->accepted_at== null || !$acceptance->declined_at==null)){
+            return redirect()->to('/hardware')->with('error', trans('admin/users/general.eula_acceptance_status'));
+        }
+        $reminder= new ReportsController();
+        $reminder->sentAssetAcceptanceReminder($acceptance->id);
```
Comment (marcusmoore) [inline app/Http/Controllers/Account/AcceptanceController.php:282]: This is going to be abstract but: calling another controller from within a controller is a code smell...

I think it would be better to encapsulate the common logic from the `sentAssetAcceptanceReminder` method and call it from both places. That encapsulation is usually in a service layer (which this app doesn't really have) or the relevant model (ie `CheckoutAcceptance@sendReminder()`).

This might not make sense so please ask any questions you have.

File: app/Http/Controllers/Account/AcceptanceController.php [hunk 3]
```diff
@@ -250,4 +238,50 @@ public function store(Request $request, $id)
...
+                                        ->get()
+                                        ->first();
+
+
+            $acceptance->accepted_at = null;
+            $acceptance->declined_at = null;
+
+            $acceptance->save();
+
+            return redirect()->to('/hardware')->with('success', trans('admin/users/general.resend_eula'));
```
Comment (marcusmoore) [inline app/Http/Controllers/Account/AcceptanceController.php:262]: The success message is displayed but I didn't get an email to resign the EULA. I think we need to send the notification after saving?

File: app/Http/Controllers/Account/AcceptanceController.php [hunk 4]
```diff
@@ -250,4 +238,50 @@ public function store(Request $request, $id)
...
+     * @return redirect
+     */
+
+    public function resignAcceptance($id){
+
+        $acceptance = CheckoutAcceptance::where('checkoutable_id', $id)
+                                        ->where('checkoutable_type','App\Models\Asset')
+                                        ->latest('created_at')
+                                        ->get()
+                                        ->first();
```
Comment (marcusmoore) [inline app/Http/Controllers/Account/AcceptanceController.php:254]: Nitpick: I think you should be be able to drop the `get()` and call `first()` directly.

**Review by marcusmoore — CHANGES_REQUESTED** [review]

**Comment by Godmartinz** [comment]
Its been a while since I last looked at this, I will refresh myself on this tomorrow and reach out with questions then.

**Comment by snipe** [comment]
Pinging you on this - can you resolve conflicts and revisit? Also @akemidx has been doing work on that assets page that I think will conflict if she beats you to the PR merge.

**Comment by akemidx** [comment]
Should be #12940 so, you can peek at that.  If you are adding a button in, we wouldn't only really have to decide where everything would end up.

tl;dr dropdown is gone, buttons added in.

**Comment by thetechjester** [comment]
@Godmartinz  I've been searching the inter-webs and this repo for this EXACT feature.

PLEASE oh PLEASE tell me this didn't get scrapped. It seems like this is something that is being largely requested in the community.

The Acceptance Form Options (especially if allowed in bulk edit) is something that is desperately needed. 
EULAs get updated/amended and currently one has to check the item out then check the item back in under the same end user in order to facilitate the Acceptance for or renewal of Acceptance and signature.

This code change/merge is functionality that is desperately needed and is way more beneficial than a button allowing one to Check In and Delete. 

How often, in the lifecycle of an asset, would one need to Check In and Delete (archive) it? The end of life or catastrophic failure. Which is typically between 3 to 6 years (depending on warranty). The EULA can be revised several times during the life cycle of an asset. Sometimes, depending on the company or technological changes it can be revised multiple times in a year.

Is this available as a php somewhere I could try out in my environment.

If not, and I'm new to posting, can this be escalated to a feature request?

Oh and since I don't see this often in posts. Just people complaining or being nasty, rude, on non-empathetic:

THANK YOU for such an awesome product and for listening to your user base! Thank you for taking the time to listen and respond courteously. Thank you for your continued development and enhancements you do.

Best!

**Comment by thetechjester** [comment]
Thank you so much @snipe !  That is wonderful news. Please, if there is a notification list I need to subscribe to in order to be alerted when this is rolled out, instruct me where to do so. I can't begin to put into words how important this feature would be in our environment. I want to make sure I know about it so I can get it in place as soon as it becomes available.

Thank you, again, for all you and the entire team do to continue making improvements to address the use-cases of your base!

You guys are the best!
Cheers!
